AX: [LiveRegionManagement] Buttons aren't included in live region announcements

hoffmanjoshua opened the pull request:
#### 5bf123876b61a6b0172fd0001980f52f929e3da0
<pre>
AX: [LiveRegionManagement] Buttons aren&apos;t included in live region announcements
<a href="https://bugs.webkit.org/show_bug.cgi?id=304289">https://bugs.webkit.org/show_bug.cgi?id=304289</a>
<a href="https://rdar.apple.com/166649406">rdar://166649406</a>

Reviewed by Tyler Wilcock.

Buttons return their inner text via `title()`, which was not in the consideration criteria
for live region objects. Now, we check if leaf elements have non-empty title before excluding
them.

Test: accessibility/mac/live-regions/live-region-button.html

* LayoutTests/accessibility/mac/live-regions/live-region-button-expected.txt: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-button.html: Added.
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::descriptionAttributeValue const):
(WebCore::AXCoreObject::title const):
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXLiveRegionManager.cpp:
(WebCore::AXLiveRegionManager::shouldIncludeInSnapshot const):

Canonical link: <a href="https://commits.webkit.org/304607@main">https://commits.webkit.org/304607@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7fa31ce619d70c74b6eb3c0d148ee0731615a17a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136003 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8356 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47278 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143703 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/88981 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/7ca2d3c3-164f-4364-a6b7-116cf863df61) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/137872 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9024 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8203 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103927 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/88981 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/0d3b7c77-ddfc-4ebd-9f3f-ecea38720c62) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138949 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6538 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121877 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84804 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/9bfd0e39-d048-40bc-8d73-fac1728a59b2) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6234 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3930 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4305 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115489 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40068 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146457 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8041 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40636 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112285 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8060 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6748 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112678 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28611 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6135 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118178 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/61983 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8089 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36243 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7810 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71646 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8031 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7889 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AXLiveRegionManager.cpp:
```diff
@@ -239,6 +239,10 @@ bool AXLiveRegionManager::shouldIncludeInSnapshot(AccessibilityObject& object) c
         return true;
 #endif
 
+    // Some leaf objects (like buttons) return their text via `title`.
+    if (!object.title().isEmpty())
+        return true;
```
`title()` and `descriptionAttributeValue()` (called above) are thin wrappers over `accessibilityText()`, which is pretty expensive to compute. Is there a not-too-ugly way to avoid duplicating that work while still accomplishing the intent of this function?

twilco replied:
What if these functions had a `Vector<AccessibilityText>*` parameter that defaults to nullptr?

e.g. `String title(Vector<AccessibilityText>* precomputedText = nullptr)`

In the function bodies, if the `precomputedText` parameter is nullptr, we can compute it in the method. I bet this optimization would apply in other places too. But if this adds too much complexity maybe it's not worth it

hoffmanjoshua replied:
Great suggestion! This is included in the latest patch. Note, we initialize the `textOrder` outside of the conditional so the pointer remains alive.

twilco commented on Source/WebCore/accessibility/AXCoreObject.h:
```diff
@@ -826,7 +826,7 @@ class AXCoreObject : public RefCountedAndCanMakeWeakPtr<AXCoreObject> {
     // This should be the visible text that's actually on the screen if possible.
     // If there's alternative text (e.g. provided by description()), that can override the title.
-    virtual String title() const;
+    virtual String title(Vector<AccessibilityText>* computedText = nullptr) const;
```
Thanks for making this change! It could be helpful to future us to have a comment on these two functions explaining the parameter a bit, e.g. something like "For performance, the computedText parameter allows passing in AccessibilityText for contexts where it has already been computed"

hoffmanjoshua committed:
AX: [LiveRegionManagement] Buttons aren't included in live region announcements
https://bugs.webkit.org/show_bug.cgi?id=304289
rdar://166649406

Reviewed by Tyler Wilcock.

Buttons return their inner text via `title()`, which was not in the consideration criteria
for live region objects. Now, we check if leaf elements have non-empty title before excluding
them.

Test: accessibility/mac/live-regions/live-region-button.html

* LayoutTests/accessibility/mac/live-regions/live-region-button-expected.txt: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-button.html: Added.
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::descriptionAttributeValue const):
(WebCore::AXCoreObject::title const):
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXLiveRegionManager.cpp:
(WebCore::AXLiveRegionManager::shouldIncludeInSnapshot const):

Canonical link: https://commits.webkit.org/304607@main
